lib/ui/window.dart

```diff
@@ -143,14 +161,262 @@ class Locale {
+  // already be canonicalized. This method does not modify parameters in any
+  // way or do any syntax checking.
+  //
+  // * language, script and region must be in normalized form.
+  // * variants must already be sorted, with each subtag in normalized form.
+  // * Iterating over variants must provide variants in alphabetical order. We
+  //   force this by taking a SplayTreeSet as parameter input, which
+  //   automatically sorts its items. We then convert to a LinkedHashSet for
+  //   faster access and iteration.
+  // * The extensions map must contain only valid key/value pairs. "u" and "t"
+  //   keys must be present, with an empty string as value, if there are any
+  //   subtags for those singletons.
+  Locale._internal(
+    String language, {
+    String script,
+    String region,
+    List<String> variants,
+    collection.SplayTreeMap<String, String> extensions,
+  }) : assert(language != null),
+       assert(language.length >= 2 && language.length <= 8),
+       assert(language.length != 4),
+       assert(script == null || script.length == 4),
+       assert(region == null || (region.length >= 2 && region.length <= 3)),
+       _languageCode = language,
+       _scriptCode = script,
+       _countryCode = region,
+       _variants = List<String>.unmodifiable(variants ?? const <String>[]),
+       _extensions = collection.LinkedHashMap<String, String>.from(extensions ?? const <String, String>{});
+
+  // TODO: unit-test this, pick the right name, make it public, update this
+  // documentation.
+  //
+  // TODO: might we prefer constructing with a single string as "variants"
+  // rather than a list of strings?
+  //
+  // TODO: this constructor modifies the list passed to variants.Is this the
+  // right design? "We mutate what you give us!" feels bad.
+  //
+  // Extensions are not yet supported by this constructor, it would require
+  // first finalizing the design of the extensions map we wish to expose.
+  factory Locale._fromComponents({
+    String language,
+    String script,
+    String region,
+    List<String> variants,
+  }) {
+    if (variants != null) {
+      variants.sort();
+    }
+    return Locale._internal(language, script: script, region: region, variants: variants);
+  }
+
+  /// Parses [Unicode CLDR Locale
+  /// Identifiers](https://www.unicode.org/reports/tr35/#Identifiers).
+  ///
+  /// This method does not parse all BCP 47 tags. See [BCP 47
+  /// Conformance](https://www.unicode.org/reports/tr35/#BCP_47_Conformance) for
+  /// details.
+  ///
+  /// TODO: support parsing all BCP 47 tags.
+  static Locale parse(String localeId) {
+    assert(localeId != null);
+    localeId = localeId.toLowerCase();
+    if (localeId == 'root')
+      return Locale._internal('und');
+
+    final List<String> localeSubtags = localeId.split(_reSep);
+    String language, script, region;
+    final List<String> variants = <String>[];
+    // Using a SplayTreeMap for its automatic key sorting.
```

Comment:
You only need the ordering to be maintained while building the class. You could switch it to a linked hash map when the map has been built (`<String, String>{}..addAll(extensions)`)

I wouldn't use `SplayTreeMap` as the type of the instenace variable. Still, it's private so it is only an implementation detail.

Reply:
I've now completely removed SplayTrees, even as temporary detail.